fs/fs_zenfs.cc

```diff
@@ -1383,9 +1385,10 @@ IOStatus ZenFS::MigrateFileExtents(
   IOStatus s = IOStatus::OK();
   Info(logger_, "MigrateFileExtents, fname: %s, extent count: %lu",
        fname.data(), migrate_exts.size());
+
   // The file may be deleted by other threads, better double check.
   auto zfile = GetFile(fname);
-  if (zfile == nullptr || zfile->IsOpenForWR()) {
+  if (zfile == nullptr || zfile->IsDeleted() || zfile->IsOpenForWR()) {
```

Comment:
What is the purpose of `zfile->IsDeleted()` here? If we are holding the `files_mtx_`, `GetFile(fname)` will return `nullptr` if the file is deleted, and `IsDeleted()` has no effect.

If we are not holding the mutex, we will potentially observe out of order behavior of `GetFile()` and `IsDeleted()`. so file will appear in the files list and is deleted may return false for a file that is in fact deleted.

Reply:
So the purpose seems to be to guard against the file being deleted or renamed after the file is selected for migration, but before migration is effected. While we do hold the mutex protecting the file map while we query the map for the file name, the lock is released when `GetFile()` returns. If the file was renamed or deleted at this point, `GetFile()` would return a `nullptr`, so no need to check `IsDeleted()`. If the file is deleted or renamed after `GetFile()` returns, this code is not guaranteed to observe the change of the `is_deleted_` field, and we may continue working with a file that is deleted.

If it is a problem that a file is deleted or renamed while migration is ongoing (which I assume, otherwise the checks are not needed), we need to synchronize on those operations, blocking until migration is done.